cpp/src/arrow/result.h

```diff
@@ -288,7 +288,7 @@ class [[nodiscard]] Result : public util::EqualityComparable<Result<T>> {
   constexpr bool ok() const { return status_.ok(); }
 
   /// \brief Equivalent to ok().
-  // operator bool() const { return ok(); }
+  constexpr explicit operator bool() const { return ok(); }
```

Comment:
Why a explicit has is added here?

Reply:
`explicit` can disable unintended implicit conversions.

Reply:
This patch LGTM, but considering https://github.com/apache/arrow/pull/4501#discussion_r293686808 . I'll let other decide whether we applying this

Reply:
The explicit way of converting to `bool` is simply calling `ok()`.

`if (result.ok()) { }`, that's all.